src/Uno.Toolkit.UI/Controls/ZoomContentControl/ZoomContentControl.cs

```diff
@@ -0,0 +1,782 @@
+	{
+		UpdateContentBoundsVisibility();
+		UpdateScrollVisibility();
+		UpdateHorizontalScrollBarValue();
+	}
+
+	private async void OnZoomLevelChanged(DependencyObject sender, DependencyProperty dp)
+	{
+		CoerceZoomLevel(sender, dp);
+		UpdateScrollLimits();
+		UpdateContentBoundsVisibility();
+		UpdateScrollVisibility();
+		await RaiseRenderedContentUpdated();
+	}
+
+	public record BoundsVisibilityData
+	{
+		public static readonly BoundsVisibilityData None = new BoundsVisibilityData(leftEdge: false, topEdge: false, rightEdge: false, bottomEdge: false);
+
+		public bool LeftEdge { get; init; }
+		public bool TopEdge { get; init; }
+		public bool RightEdge { get; init; }
+		public bool BottomEdge { get; init; }
+		public bool AllEdgesVisible { get; init; }
+
+		public BoundsVisibilityData(bool leftEdge, bool topEdge, bool rightEdge, bool bottomEdge)
+		{
+			LeftEdge = leftEdge;
+			TopEdge = topEdge;
+			RightEdge = rightEdge;
+			BottomEdge = bottomEdge;
+			AllEdgesVisible = LeftEdge && TopEdge && RightEdge && BottomEdge;
+		}
+	}
+
+	private void UpdateContentBoundsVisibility()
+	{
+		if (_presenter?.Content is FrameworkElement fe)
+		{
+			var m = GetPositionMatrix(fe, this);
+
+			ContentBoundsVisibility = new BoundsVisibilityData(
+				m.OffsetX >= 0,
+				m.OffsetY >= 0,
+				this.ActualWidth >= (fe.ActualWidth * ZoomLevel) + m.OffsetX,
+				this.ActualHeight >= (fe.ActualHeight * ZoomLevel) + m.OffsetY);
+		}
+	}
+
+	private void UpdateScrollVisibility()
+	{
+		IsHorizontalScrollBarVisible = !(ContentBoundsVisibility.RightEdge && ContentBoundsVisibility.LeftEdge);
+		IsVerticalScrollBarVisible = !(ContentBoundsVisibility.TopEdge && ContentBoundsVisibility.BottomEdge);
+	}
+
+	private bool CanMoveIn((bool Horizontal, bool Vertical) _movementDirection)
+	{
+		if (ContentBoundsVisibility.AllEdgesVisible)
+		{
+			return false;
+		}
+
+		var canMove = false;
+		canMove |= CanScrollLeft() && _movementDirection.Horizontal is true;
+		canMove |= CanScrollRight() && _movementDirection.Horizontal is false;
+		canMove |= CanScrollUp() && _movementDirection.Vertical is true;
+		canMove |= CanScrollDown() && _movementDirection.Vertical is false;
+
+		return canMove;
+	}
+
+	private bool CanScrollUp() => !ContentBoundsVisibility.TopEdge;
+
+	private bool CanScrollDown() => !ContentBoundsVisibility.BottomEdge;
+
+	private bool CanScrollLeft() => !ContentBoundsVisibility.LeftEdge;
+
+	private bool CanScrollRight() => !ContentBoundsVisibility.RightEdge;
+
+	//Slide move is always on the opposite direction of the drag
+	private async void UpdateVerticalScrollBarValue()
+	{
+		VerticalScrollValue = VerticalOffset;
+		await RaiseRenderedContentUpdated();
+	}
+
+	//Slide move is always on the opposite direction of the drag
+	private async void UpdateHorizontalScrollBarValue()
+	{
+		HorizontalScrollValue = -1 * HorizontalOffset;
+		await RaiseRenderedContentUpdated();
+	}
```

Comment:
if it is "on the opposite direction of the drag", where is the -1 multiplier on the vertical offset?

Reply:
In the screen, y axis is already inverted, i.e. the Y=0 is the top line.

Reply:
maybe we could remove the comment for Y-axis/Vertical, or clarify that in the comment
it is otherwise confusing to see "opposite" but not using the -1 multiplier